Some cleanings in the views (dashboard, layout, wall)

mohierf opened the pull request:
And cleanings in the hosts/services synthesis computation:
- say that an host/service is a problem when HARD non OK state
- make this consistent on all views (dashboard, currently, popovers, problems)

Manage a configurable minimum business impact
- set defaut to 2
- badges in the header bar only for problems > minimum BI

This has been tested in a test environment. Perharps some issues will raise in some specific case  ... but I will fix rapidly 😉

mohierf committed:
Some cleanings in the views (dashboard, layout, wall)
And cleanings in the hosts/services synthesis computation
Minimum business impact

maethor commented on module/datamanager.py:
```diff
@@ -1000,21 +1024,42 @@ def get_overall_it_state(self, user):
         if "is:impact" not in search:
             search = "is:impact " + search
+        if "bi:" not in search:
+            search = "bi:>=%d " % self.min_business_impact + search
```
Please put parenthesis around the % operation, for readability.

mohierf replied:
Ok

maethor commented on module/datamanager.py:
```diff
@@ -1000,21 +1024,42 @@ def get_overall_it_state(self, user):
-                     get_acknowledged=False, get_downtimed=False, sorter=worse_first):
+    def get_problems(self, user, search='isnot:UP isnot:OK isnot:PENDING isnot:UNKNOWN isnot:ACK isnot:DOWNTIME '
+                                        'is:HARD type:all', get_acknowledged=False, get_downtimed=False,
+                     sorter=worse_first):
```
I disagree. UNKNOWN IS a problem. You can get an UNKNOWN when a check plugin doesn't work, if you filter on this you will never see that the plugin is not working (and the service may not be OK). Worse, if you don't configure thresholds properly, you can get an UNKNOWN on `check_http` when the site timeout. What does it mean ? Is it not a problem ?

I cannot merge that. Please don't change this query.

maethor replied:
Please do not remove the bi filter. As discussed by email, you need to differentiate "min_business_impact" for the top badges and for the default problems query.

In my case, I want to keep a look on the "low importance" problems, like pending upgrades or problems on staging environments. But I don't want this problems to be counted in the top badges, because they do not need immediate attention or fixing.

mohierf replied:
It was a suggestion.

1. for UNKNOWN, it is a matter of view ... I can accept your analysis. I proposed to exclude UNKNOWN because Nagios does ... but I agree with you. I will revert this

1. for the BI filter I also agree. I propose another approachh ... each contact in the configuration has a `min_business_impact`  in its attributes and we should use this as a default filte for the problems view. The default value is 0 to view all the problems but if you set 2 for the user of the webui you will have a de-facto filter. It would make sense ...

maethor replied:
About point 2. it makes sense. But I think you still need to have two variables, one for problems view and one for the badge.

mohierf replied:
Hmmm... I will also check if the datamanger considers the current user `min_bi`. And then we will have this order in the filtering:
 - user `min_bi` filtering first
 - main bi filtering for the whole UI (including problems view)
 - badge bi filtering for the badges

maethor commented on module/datamanager.py:
```diff
@@ -1000,21 +1024,42 @@ def get_overall_it_state(self, user):
+        if "is:HARD" not in search:
+            search = "is:HARD" + search
+        if "bi:" not in search:
+            search = "bi:>=%d " % self.min_business_impact + search
```
NO! Why are you doing all this!?! No!

mohierf replied:
The*get_problems* function is only used in the Wall view and it needs to be consistent with all the UI ... why forcing only 3 criteria whereas more are necessary to qualify a problem ...

maethor replied:
OK.

maethor commented on module/views/header_element.tpl:
```diff
@@ -37,7 +37,7 @@
        <table class="table table-invisible table-condensed">
          <tbody>
             <tr>
-               %for state in "up", "unreachable", "down", "pending", "unknown", "ack", "downtime":
```
Why did you removed pending?

mohierf replied:
Pending is only a transient state that is not not of a great interest in the pop over ...  this state only existswhen a check has not yet been launched by Shinken. If a host/service has this state it means that your Shinken stopped just after the WebUI staterd :/

maethor replied:
OK.

maethor commented on module/datamanager.py:
```diff
@@ -121,17 +122,15 @@ def get_percentage_hosts_state(self, user=None, problem=False):
-        problems = [h for h in all_hosts if h.is_problem and not h.is_impact]
+        known_problems = h['nb_problems']
+        if not problem:
+            known_problems = h['nb_elts'] - h['nb_problems']
```
`not problem` is default in the function definition. So please begin by `if not problem: … else: …`

mohierf replied:
Ok. I will change the test

maethor commented:
I'm sorry I cannot accept this PR. I leave you my comments in the review.

mohierf committed:
Some updates after the 1st review

Coloured badges in the navigation bar
User minimum business impact
UI minimum business impact
Badge minimum business impact
UI own problems computation algorithm

mohierf commented:
No problem that you did not accepted, it is the interest of the review 😉 and I appreciate your comments

mohierf commented:
Concerning the BI filters, the user `min_business_impact` filters the default UI ...

For the same configuration, the *admin* user is default (0)
---
![image](https://user-images.githubusercontent.com/2932687/49665386-ffc38280-fa54-11e8-92ec-9134f477734c.png)

whereas the *guest* user is important (3)
---

![image](https://user-images.githubusercontent.com/2932687/49665318-ce4ab700-fa54-11e8-9794-f436a68b7657.png)

mohierf commented:
The badge filtering when set to Very important (4)
![image](https://user-images.githubusercontent.com/2932687/49667727-2e912700-fa5c-11e8-94c0-1d0ebba52d07.png)

Or set to normal (2):
![image](https://user-images.githubusercontent.com/2932687/49667771-541e3080-fa5c-11e8-9fc5-675d908c35e1.png)

mohierf commented:
Hosts / services global status remain like all the navbar icon, but the badge (if present...) is colored warning or critical:
![image](https://user-images.githubusercontent.com/2932687/49687259-e11ac580-fb00-11e8-919e-12748fa5af1a.png)


Looks better 😉

mohierf commented:
I rebased on the most recent develop branch. Should be merged now 😉

maethor commented on module/datamanager.py:
```diff
@@ -1000,21 +1053,41 @@ def get_overall_it_state(self, user):
+        """
         if "isnot:UP" not in search:
-            search = "isnot:UP " + search
+            search = search + " isnot:UP"
```
Why did you change this? It is way better to filter first on "isnot:UP" because it is one of the fastest and more discriminent filter.

mohierf replied:
+1 for you! I did a bad copy/paste ... I fix this and I will rebase/squash the commits for you merge this

maethor commented on module/module.py:
```diff
@@ -341,6 +344,18 @@ def __init__(self, modconf):
+        # --------------------------
+        # All the hosts and services that are in a HARD non OK/UP state are considered as problems if their
+        # business_impact is greater than or equal this value
+        self.min_business_impact = int(getattr(modconf, 'min_business_impact', '0'))
```
Should be named `problems_min_business_impact` or maybe `problems_business_impact`.

maethor replied:
Should default to 1 for consistent behavior, and because 0 officially means "no-importance" in Shinken docs.

mohierf replied:
Ok for renaming and default value

maethor commented on module/module.py:
```diff
@@ -341,6 +344,18 @@ def __init__(self, modconf):
+        # business_impact is greater than or equal this value
+        self.min_business_impact = int(getattr(modconf, 'min_business_impact', '0'))
+        # badge_business_impact is used to filter the alerting badges in the header bar (default is 2)
+        self.badge_business_impact = int(getattr(modconf, 'badge_business_impact', '2'))
```
Should be named `important_problems_min_business_impact` or maybe `important_problems_business_impact`.

These variables should be named with "user meaning" in mind, not technical stuff or visual element like "badge". Because the point is to determinate which alert is a problem, and which is an important problem. Once you have this info, you use it anyway you need. Today it is for the badge in the header bar, but tomorrow it may be somewhere else.

mohierf replied:
Ok. I agree wit this. I will rename as `important_problems_business_impact`

maethor commented on module/module.py:
```diff
@@ -341,6 +344,18 @@ def __init__(self, modconf):
+                    self.min_business_impact, self.badge_business_impact)
+
+        # Inner computation rules for the problems
+        self.inner_problems_count = int(getattr(modconf, 'inner_problems_count', '1'))
```
Why do you leave this option to user? Can't we just abandon the old way and adopt the new one if it is better?

If you need to keep this option, please change it to `enable_inner_problems_count`. Because `inner_problems_count` evokes a counter, not a boolean.

mohierf replied:
I keep this option configurable because it may be interesting to have a view on the real computation done by the framewrok. 
I will rename as `disable_inner_problems_computation`

maethor commented on module/plugins/problems/problems.py:
```diff
@@ -40,6 +40,10 @@ def get_page():
+
+    # Update the default filter according to the logged-in user minimum business impact
+    default_filtering = DEFAULT_FILTER + " bi:>=%d" % (user.min_business_impact)
+
```
May not work if `user.min_business_impact` is lesser than `app.min_business_impact`. I didn't check.

maethor replied:
Please be consistent in variable names. You can name this variable `default_filter` instead of `default_filtering`. By using lowercase instead of caps we will understand it is a local version of the variable.

mohierf replied:
Note that `default_filtering`is only a local variable ... and this is only used to set a bollean value for the vie template. No search is executed with this criterai ... only set True/False for the empty search message in the problems view.

No modification in the code for this comment

maethor commented on module/views/_filters.tpl:
```diff
@@ -24,7 +24,8 @@
               <li role="presentation" class="divider"></li>
-              <li role="presentation"><a role="menuitem" href="/all?search=isnot:0 isnot:ack isnot:downtime&title=New problems">New problems</a></li>
+              <li role="presentation"><a role="menuitem" href="/all?search=isnot:0 is:hard isnot:ack isnot:downtime&title=New problems">New problems</a></li>
+              <li role="presentation"><a role="menuitem" href="/all?search=isnot:0 is:soft isnot:ack isnot:downtime&title=Potential problems">Potential problems</a></li>
```
So in "New problems" you are displaying PENDING, but not SOFT? Besides this, "New problems" is the same search than /problems. I don't see the point of keeping it.

mohierf replied:
Yes.
**Nxew problems** are problems not yet managed (HARD not OK/UP which are not yet ack/downtime)
**Potential problems** are the same but not yes confirmed (HARD) ...

maethor commented on module/views/header_element.tpl:
```diff
@@ -33,11 +33,14 @@
      %h = app.datamgr.get_hosts_synthesis(user=user)
+     %# Override the default problems count - badge minimum business impact
+     %h['nb_problems'] = len(app.datamgr.get_problems(user,'type:host bi:>%d' % app.badge_business_impact))
+     %s['nb_problems'] = len(app.datamgr.get_problems(user,'type:service bi:>%d' % app.badge_business_impact))
```
`user.min_business_impact` should be used if it is greater than `badge_business_impact`.

`max(user.min_business_impact, app.badge_business_impact)` may do the trick.

Better, I think the datamanager should store the `badge_business_impact` (`important_problems_business_impact`) variable, and use directly the `user.min_business_impact` variable. Here you should just need to call `app.datamgr.get_important_problems(user, 'type:host')`.

Even better, why do you need to add this in synthesis. Why `get_hosts_synthesis` and `get_services_synthesis` did not already compute the right count?

mohierf replied:
I agree with your comment ... obviously, the best is refactoring of the synthesis functions to integrate all this stuff ... using a simple boolean parameter for important or not proiblems 😉 

 Ipropose to log an issue for this as a future enhancement

maethor commented:
@mohierf OK. I review everything one more time. I leave some comments. Once this is done I will merge this PR.

mohierf committed:
Some updates after the 2nd review

mohierf committed:
Some updates after the 3rd review
